### PR TITLE
Xdmf io in `Idefix`

### PR DESCRIPTION
**Pull request by dutta-alankar**
I have added `Xdmf` IO capability in `Idefix` using a header-only `hdf5` package `HighFive`. I'm new to all of these, name C++, Kokkos and Idefix. I made the `hdf5` (and companion `xmf`) files as similar as possible to the `PLUTO` generated outputs.
Please review and suggest whatever changes you deem fit in here.

**TODO**: 
- Need to implement `hdf5` chunking to make the io faster. Currently, loads of the data dumps are slow.
- Add compression filters to reduce file size on disk.
- Sometimes with some number of processors (say 8 but not with 4) running in parallel, the program aborts with a segmentation fault (signal 11) during the second hdf5 data dump. However, the initial data dump has no issues. This needs investigation and bug-fixed.
- Need to test and confirm if the Z-Y-X dumping order is consistently maintained everywhere by running the dumps for non-equal grid points in all directions. Also need to test the dumps in 2D and 1D.

### CODE REVIEW

File: src/output/output.hpp [hunk 1]
```diff
@@ -13,6 +13,7 @@
 #include "input.hpp"
 #include "dataBlock.hpp"
 #include "vtk.hpp"
+#include "xdmf.hpp"
```
Comment (dutta-alankar) [inline src/output/output.hpp:16]: This needs to be included only when needed!

File: src/output/output.hpp [hunk 2]
```diff
@@ -25,13 +26,15 @@ using UserDefVariablesFunc = void (*) (DataBlock &, UserDefVariablesContainer &)
...
   friend class Vtk;     // Allow VTK to have access to user-defined variables
+  friend class Xdmf;    // Allow XDMF to have access to user-defined variables
   friend class DumpImage; // Allow dumpimag to have access to dump API
  public:
   Output(Input &, DataBlock &);           // Create Output Object
   int CheckForWrites(DataBlock &);        // Check if outputs are needed at this stage
   void RestartFromDump(DataBlock &, int);  // Restart from a dump file.
   void ForceWriteDump(DataBlock &);            // Force write dumps (needed during an abort)
   void ForceWriteVtk(DataBlock &);            // Force write vtks
+  void ForceWriteXdmf(DataBlock &);          // Force write xdmfs
```
Comment (dutta-alankar) [inline src/output/output.hpp:37]: Needs to be included on demand.

File: src/output/output.hpp [hunk 3]
```diff
@@ -40,6 +43,7 @@ class Output {
  private:
   Vtk vtk;          // local instance of Vtk class
   Dump dump;        // local instance of Dump class
+  Xdmf xdmf;        // local instance of Xdmf class
```
Comment (dutta-alankar) [inline src/output/output.hpp:46]: Same. To be included on demand. Not by default.

Comment (dutta-alankar) [reply src/output/output.hpp:46]: @glesur @neutrinoceros Can you suggest a way to do this?

**Comment by neutrinoceros** [comment]
@dutta-alankar Can you explain how vtk and dump files are not suited for however you want to analyse/post-process your results ?

The decision to include this or a variant of it is ultimately @glesur's, but FYI we are currently working on Idefix v2.0 (in a private repo), which, among other things, is a pretty significant refactor of how output routines are implemented, so I don't except that we'll add support for new output formats within idefix 1.x

**Comment by dutta-alankar** [comment]
@neutrinoceros It is extremely convenient to load raw data stored in hdf5 files in python and use them for different analyses. I also have a lot of analysis codes that use the PLUTO hdf5 for analysis and these can be used without much porting on Idefix. For example, the following is a portion taken from a code I recently used for the analysis of some `PLUTO` hdf5 files.

Besides this, I also have developed an In-Situ visualization plugin that works with `PLUTO` using `ParaView Catalyst`. I can happily put that in as well in `Idefix` in either v1.0 or v2.0 if you or @glesur are interested.

```python
import numpy as np
import h5py

for snap in range(tot_snap):
    print(snap, end='\r')
    step = 10*snap
    tcc[snap] = step/np.sqrt(chi)
    with h5py.File(f'{snap_dir}/data.{snap:04d}.flt.h5', 'r') as data:
        dV = np.array(data[f'/Timestep_{snap}/vars/cellvol']).flatten()
        temperature = np.array(data[f'/Timestep_{snap}/vars/temperature']).flatten()
        ndens = np.array(data[f'/Timestep_{snap}/vars/ndens']).flatten()
        pres = np.array(data[f'/Timestep_{snap}/vars/pressure']).flatten()
        tracer = np.array(data[f'/Timestep_{snap}/vars/tr1']).flatten()

        x = np.array(data['/cell_coords/X']).flatten()
        y = np.array(data['/cell_coords/Y']).flatten()
        z = np.array(data['/cell_coords/Z']).flatten()

        distance = np.sqrt(x**2 + y**2 + z**2)
        rad_pnts = 100
        radius = np.linspace(0.98*np.min(distance), 1.02*np.max(distance), rad_pnts+1)

        cond_cold_gas = temperature<(3.3*Tcl)
```

**Comment by glesur** [comment]
Thanks @dutta-alankar for your effort in adding hdf5 to idefix. I believe it's a good idea on the long term as vtk is getting old and will not be sufficient when we will start developing an AMR layer. However, I would like idefix to remain as library-free as possible for the moment, so hdf5 should come as an optional feature, that is not required at compile time by default. I believe this is also true for Pluto.

**Comment by dutta-alankar** [comment]
@glesur Thanks for taking time to have a look at this PR. Previously using HighFive the parallel dump wasn't working. So I switched to using native HDF5 functions. Without using added library layer the io time has also reduced significantly.

**Comment by dutta-alankar** [comment]
Also, wanted to add, just like PLUTO, I have added hdf5 as an optional dependency keeping in line with the view that default Idefix is as much library independent and stand-alone as possible.

File: src/output/output.hpp (hunk 2 above)
Comment (glesur) [reply src/output/output.hpp:37]: We probably need a master #define flag to enable all of this machinery from Cmake. something like `Idefix_XDMF_OUTPUT`

File: src/output/output.hpp (hunk 3 above)
Comment (dutta-alankar) [reply src/output/output.hpp:46]: @glesur I have already put this to be optionally included if the pre-processor flag `USE_HDF5` is defined in `CMakeLists.txt` in the root directory of Idefix.

**Comment by glesur** [comment]
@dutta-alankar you need to secure all the `#include "hdf5.hpp"` by a guard `#ifdef USE_HDF5 / #endif` so that the code compile without hdf5 installed.

**Comment by dutta-alankar** [comment]
> @dutta-alankar you need to secure all the `#include "hdf5.hpp"` by a guard `#ifdef USE_HDF5 / #endif` so that the code compile without hdf5 installed.

Fixed it!

**Comment by dutta-alankar** [comment]
Some more places needed fixing!

**Comment by glesur** [comment]
The xdmf write function seem to work fine (tested 4 and 8 procs with various domain decomposition), but the read always lead to segfaults on my laptop. I think this should be fixed before being merged.

Also, you should use the standard `IDEFIX_ERROR` macro when xdmf is requested by the user and not linked against idefix.

**Comment by dutta-alankar** [comment]
@glesur Can you please elaborate on the error that you encounter? This will help me write a fix. I have tried running Idefix on a test problem which is SedovBlast and it seems to be working and giving hdf5 files and companion xmf files. I loaded them on ParaView and it seems okay.

File: .gitmodules
Comment (glesur) [inline .gitmodules:1]: This should be made optional, we don't want to pull highfive on a systematic basis.

File: .gitmodules [hunk 4]
```diff
@@ -1,3 +1,6 @@
 [submodule "src/kokkos"]
     path = src/kokkos
     url = https://github.com/kokkos/kokkos.git
+[submodule "src/HighFive"]
```
Comment (glesur) [inline .gitmodules:4]: This should be made optional, we don't want to pull highfive on a systematic basis.

File: src/output/output.cpp [hunk 5]
```diff
@@ -184,6 +231,29 @@ void Output::ForceWriteVtk(DataBlock &data) {
   idfx::popRegion();
 }
 
+#ifdef USE_HDF5
```
Comment (glesur) [inline src/output/output.cpp:234]: to be consistent with MPI, this should probably be `#ifdef WITH_HDF5`, and same for all of the occurences of this flag.

**Review by glesur — CHANGES_REQUESTED** [review]
a few minor things to guarantee consistancy with previous #ifdef flag, and make sure that the highfive submodule stays optional.

File: .gitmodules (hunk 4 above)
Comment (dutta-alankar) [reply .gitmodules:4]: I initially was using the HighFive API for HDF5 dumps but encountered some problems with parallel io and also the io times were karge due to overhead. I rewrote the entire thing to native HDF5 API and so I don't need HighFive anymore. I forgot to remove this from `.gitmodules`. Getting rid of this.

**Review by glesur — APPROVED** [review]

**Comment by dutta-alankar** [comment]
Can this PR be merged now? Please let me if any more changes are needed.

**Pull request merged**